### PR TITLE
Use fetched data from API to create all keyword cards on page

### LINKED ISSUE #8: Change the 'generateCard' script to use database
The preliminary script used for creating cards needs to be changed to pull the keyword names and description from our database.

### PR DESCRIPTION
**Pull request by OscarLarm**
### Description
The generateCard.js script has been changed to fetch all keywords and descriptions from the API. Now every keyword in the database should have its own card created and rendered on the page. 
The script stores each keyword property in one array and every matching description property in another array.
I have also added try and catch statements for error handling.

### Details
- Get all card data from API
- Added error handling for generateCard.js

### References
Close #8 

<br>

### CODE REVIEW

File: frontend/generateCard.js [hunk 1]
```diff
@@ -1,65 +1,74 @@
...
+            throw new Error(`HTTP error! status: ${response.status}`);
+        }
+        const json = await response.json();
+        cardsData = json; //Full api data
 
-// Function to create a card element with a title and description
-function makeCard(title, description) {
-    // Create a new div element for the card
+        // Taking the keyword titles and descriptions
+        cardsData.forEach(card => {
```
Comment (annk15) [inline frontend/generateCard.js:20]: I don't think we need to extract the attributes from the parsed json object here. We should remove the arrays and move the cardsData variable out from the method and access the attributes using a for each loop when constructing the cards. Then we can remove some of the logic and reduce the complexity and optimise it a bit as no data need to be copied.

File: frontend/generateCard.js [hunk 2]
```diff
@@ -1,65 +1,74 @@
...
-function generateCards() {
-    for (let i = 0; i < titles.length; i++) {
-        const cards = makeCard(titles[i], descriptions[i]);
-        cardHolder.appendChild(cards);
+    return card; // Return the constructed card element
+}
+// Renders all the cards on the page
+async function renderCards() {
+    try {
+        for (let i = 0; i < keywordData.length; i++) {
```
Comment (annk15) [inline frontend/generateCard.js:57]: This should be a foreach loop and call the attributes directly on the cardsData object as it contains all the information needed.

File: frontend/generateCard.js [hunk 3]
```diff
@@ -1,65 +1,74 @@
...
-    "Reach (This creature can block creatures with flying.)",
-    "Hexproof (This permanent can't be the target of spells or abilities your opponents control.)",
-    "Double strike (This creature deals both first-strike and regular combat damage.)",
-    "First strike (This creature deals combat damage before creatures without first strike.)",
-    "Lifelink (Damage dealt by this creature also causes you to gain that much life.)"
-];
+    try {
+        const response = await fetch(url);
+        if (!response.ok) {
+            throw new Error(`HTTP error! status: ${response.status}`);
```
Comment (annk15) [inline frontend/generateCard.js:14]:
```suggestion
            throw new Error(`Error : Endpoint (Keyword abilities) responded with: ${response.status}`);
```

File: frontend/generateCard.js [hunk 4]
```diff
@@ -1,65 +1,74 @@
...
-    'Haste',
-    'Flying',
-    'Reach',
-    'Hexproof',
-    'Double Strike',
-    'First Strike',
-    'Lifelink'
-];
+// Fetch card data from API
+async function fetchCard() {
```
Comment (annk15) [inline frontend/generateCard.js:8]: Should this be called "fetchCardData" instead?

File: frontend/generateCard.js [hunk 5]
```diff
@@ -1,65 +1,74 @@
...
-// Function to generate and append cards based on titles and descriptions arrays
-function generateCards() {
-    for (let i = 0; i < titles.length; i++) {
-        const cards = makeCard(titles[i], descriptions[i]);
-        cardHolder.appendChild(cards);
+    return card; // Return the constructed card element
+}
+// Renders all the cards on the page
+async function renderCards() {
+    try {
```
Comment (annk15) [inline frontend/generateCard.js:56]: I don't think we need this outer try catch. We can just check if keywordData is null or undefined with a if statement as that is the only type of error that can occur.

File: frontend/generateCard.js [hunk 6]
```diff
@@ -1,65 +1,74 @@
...
+        for (let i = 0; i < keywordData.length; i++) {
+            try {
+                const card = createCard(keywordData[i], descriptionData[i]);
+                cardHolder.appendChild(card);
+            } catch (innerError) {
+                console.log('Error rendering card:', innerError.message);
```
Comment (annk15) [inline frontend/generateCard.js:62]: If we replace the outer for loop with a foreach we can print the attributes in the log of the card which failed to make it easier to debug.

**Review by annk15 — CHANGES_REQUESTED** [review]
I think this works as expected but we should try to optimise it a bit more to make it easier to maintain. I have left some suggestions on improvements, just reply to the comment if you have any questions or don't agree and we can have a discussion about it :shipit:

File: frontend/generateCard.js (hunk 4 above)
Comment (OscarLarm) [reply frontend/generateCard.js:8]: Good idea

File: frontend/generateCard.js (hunk 1 above)
Comment (OscarLarm) [reply frontend/generateCard.js:20]: Now the fetchCardData fetches data directly without putting it into a variable, returning the extracted .json object directly.
The cardData variable gets declared and defined with the fetch data in the renderCards() function, then creates and renders a card forEach object in cardData

**Comment by OscarLarm** [comment]
> I think this works as expected but we should try to optimise it a bit more to make it easier to maintain. I have left some suggestions on improvements, just reply to the comment if you have any questions or don't agree and we can have a discussion about it :shipit:

All changes has been implemented, but I'm unable to push them. Should I make a new branch called feature/cards-from-api and make a new pull request?

**Comment by annk15** [comment]
> > I think this works as expected but we should try to optimise it a bit more to make it easier to maintain. I have left some suggestions on improvements, just reply to the comment if you have any questions or don't agree and we can have a discussion about it :shipit:
> 
> All changes has been implemented, but I'm unable to push them. Should I make a new branch called feature/cards-from-api and make a new pull request?

yes sorry about that, I have setup a rule so we need to prefix all branches with feature/.
You need as you said create a new branch and PR, the branch should be named "feature/cards-from-api" 👍 
You can checkout development, make a pull and then merge this branch in to that one. You will then get a conflict that needs to be resolved before you can push, you should always use the code from development if the conflict occurs in any of my files.

**Pull request closed without merging**